File: backend/src/main/java/com/nexacore/examenes/security/PasswordPolicy.java
package com.nexacore.examenes.security;

import java.nio.charset.StandardCharsets;

/** Requisitos para contraseñas elegidas por el usuario, no para claves temporales. */
public final class PasswordPolicy {

    private PasswordPolicy() {}

    public static void validate(String password) {
        if (password == null || password.isBlank()) {
            throw new IllegalArgumentException("La nueva contraseña es obligatoria");
        }
        if (password.codePointCount(0, password.length()) < 8) {
            throw new IllegalArgumentException("La nueva contraseña debe tener al menos 8 caracteres");
        }
        if (password.getBytes(StandardCharsets.UTF_8).length > 72) {
            throw new IllegalArgumentException("La nueva contraseña es demasiado larga");
        }
        if (password.codePoints().noneMatch(Character::isUpperCase)) {
            throw new IllegalArgumentException("La nueva contraseña debe incluir una mayúscula");
        }
        if (password.codePoints().noneMatch(Character::isLowerCase)) {
            throw new IllegalArgumentException("La nueva contraseña debe incluir una minúscula");
        }
        if (password.codePoints().noneMatch(Character::isDigit)) {
            throw new IllegalArgumentException("La nueva contraseña debe incluir un número");
        }
        if (password.codePoints().noneMatch(PasswordPolicy::isSymbol)) {
            throw new IllegalArgumentException("La nueva contraseña debe incluir un símbolo");
        }
    }

    private static boolean isSymbol(int codePoint) {
        int type = Character.getType(codePoint);
        return type == Character.CONNECTOR_PUNCTUATION
                || type == Character.DASH_PUNCTUATION
                || type == Character.START_PUNCTUATION
                || type == Character.END_PUNCTUATION
                || type == Character.INITIAL_QUOTE_PUNCTUATION
                || type == Character.FINAL_QUOTE_PUNCTUATION
                || type == Character.OTHER_PUNCTUATION
                || type == Character.MATH_SYMBOL
                || type == Character.CURRENCY_SYMBOL
                || type == Character.MODIFIER_SYMBOL
                || type == Character.OTHER_SYMBOL;
    }
}

File: frontend/src/utils/passwordPolicy.ts
export const PASSWORD_REQUIREMENTS =
  'Mínimo 8 caracteres; incluye mayúscula, minúscula, número y símbolo.'

/** Política de contraseñas nuevas, compartida por perfil y restablecimiento. */
export function validateNewPassword(value: string): string {
  if (!value.trim()) return 'La nueva contraseña es obligatoria'
  if ([...value].length < 8) return 'Mínimo 8 caracteres'
  if (new TextEncoder().encode(value).length > 72) return 'La contraseña es demasiado larga'
  if (!/\p{Lu}/u.test(value)) return 'Incluye al menos una mayúscula'
  if (!/\p{Ll}/u.test(value)) return 'Incluye al menos una minúscula'
  if (!/\p{Nd}/u.test(value)) return 'Incluye al menos un número'
  if (!/[\p{P}\p{S}]/u.test(value)) return 'Incluye al menos un símbolo'
  return ''
}
